### PR TITLE
[Merged by Bors] - feat: define the derived set

### PR DESCRIPTION
**Pull request by Command-Master**
Define the derived set of a set, and prove some properties. Also prove `AccPt.map`.

---

- [x] depends on: #14787

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/DerivedSet.lean [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  intro U hu1 hu2
+  obtain ⟨y, hy, ⟨hy1, hy2⟩⟩ := hx U hu1 hu2
+  rw [derivedSet, Set.mem_setOf, accPt_iff_frequently, frequently_nhds_iff] at hy2
+  obtain ⟨z, hz⟩ := hy2 (U \ {x}) (IsOpen.sdiff hu1 isClosed_singleton) (by simp [hy1, hy])
+  simp at hz
+  use z
+  simp [hz]
+
+lemma IsPreconnected.subset_derivedSet_self [T1Space X] {U : Set X} (hu : U.Nontrivial)
+    (h : IsPreconnected U) : U ⊆ derivedSet U := by
```
Comment (Command-Master) [inline Mathlib/Topology/DerivedSet.lean:63]: Is there a better spelling for a set being dense-in-itself?

Comment (Command-Master) [reply Mathlib/Topology/DerivedSet.lean]: There's Preperfect, I changed it to use it.

**Comment by j-loreaux** [comment]
Can you please merge master?

**Comment by Command-Master** [comment]
I see the changes of #14787 were duplicated, what caused this?

**Comment by j-loreaux** [comment]
They're not duplicated, but the diff in this PR was against a version of Master prior to when your other PR was merged.

File: Mathlib/Topology/DerivedSet.lean [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  apply map_neBot (m := f) (hf := h) |>.mono
+  rw [Filter.map_inf hf2]
+  gcongr
+  apply tendsto_nhdsWithin_of_tendsto_nhds_of_eventually_within _ hf1.continuousWithinAt
+  simpa [hf2.eq_iff] using eventually_mem_nhdsWithin
+
+/--
+The derived set of a set is the set of all accumulation points of it.
+-/
+def derivedSet (A : Set X) : Set X := {x | AccPt x (𝓟 A)}
```
Comment (j-loreaux) [inline Mathlib/Topology/DerivedSet.lean:32]: I wonder if there's any reason *not* to allow for arbitrary filters as inputs. I guess it means you can't iterate the `derivedSet` as easily.

Comment (Command-Master) [reply Mathlib/Topology/DerivedSet.lean:32]: The theorems would be less pretty, and if you're using an arbitrary filter you could just directly use `AccPt` — how often do you want it as a set?

File: Mathlib/Topology/DerivedSet.lean [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  obtain ⟨y, hy1, rfl⟩ := hx
+  convert hy1.map hf1.continuousAt hf2
+  simp
+
+lemma isClosed_iff_derivedSet_subset (A : Set X) : IsClosed A ↔ derivedSet A ⊆ A where
+  mp h := by
+    rw [isClosed_iff_clusterPt] at h
+    intro x hx
+    apply h
+    apply hx.clusterPt
```
Comment (j-loreaux) [inline Mathlib/Topology/DerivedSet.lean:58]: With this lemma:
```lean
lemma derivedSet_subset_closure (A : Set X) : derivedSet A ⊆ closure A :=
  fun _ hx ↦ mem_closure_iff_clusterPt.mpr hx.clusterPt
```
you can golf the `mp` direction to:
```suggestion
lemma isClosed_iff_derivedSet_subset (A : Set X) : IsClosed A ↔ derivedSet A ⊆ A where
  mp h := derivedSet_subset_closure A |>.trans h.closure_subset
```

File: Mathlib/Topology/DerivedSet.lean [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  rw [isClosed_iff_derivedSet_subset]
+  intro x hx
+  rw [derivedSet, Set.mem_setOf, accPt_iff_frequently, frequently_nhds_iff] at hx ⊢
+  intro U hu1 hu2
+  obtain ⟨y, hy, ⟨hy1, hy2⟩⟩ := hx U hu1 hu2
+  rw [derivedSet, Set.mem_setOf, accPt_iff_frequently, frequently_nhds_iff] at hy2
+  obtain ⟨z, hz⟩ := hy2 (U \ {x}) (by simp [hy1, hy]) (IsOpen.sdiff hu2 isClosed_singleton)
+  simp at hz
+  use z
+  simp [hz]
```
Comment (j-loreaux) [inline Mathlib/Topology/DerivedSet.lean:78]: You can prove (with `import Mathlib.Tactic.Peel`):
```lean
/-- In a `T1Space`, the `derivedSet` of the closure of a set is equal to the derived set of the
set itself.

Note: this doesn't hold in a space with the indiscrete topology. For example, if `X` is a type with
two elements, `x` and `y`, and `A := {x}`, then `closure A = Set.univ` and `derivedSet A = {y}`,
but `derivedSet Set.univ = Set.univ`. -/
lemma derivedSet_closure [T1Space X] (A : Set X) : derivedSet (closure A) = derivedSet A := by
  refine le_antisymm (fun x hx => ?_) (derivedSet_mono _ _ subset_closure)
  rw [mem_derivedSet, AccPt, (nhdsWithin_basis_open x {x}ᶜ).inf_principal_neBot_iff] at hx ⊢
  peel hx with u hu _
  obtain ⟨-, hu_open⟩ := hu
  exact mem_closure_iff.mp this.some_mem.2 (u ∩ {x}ᶜ) (hu_open.inter isOpen_compl_singleton)
    this.some_mem.1
```
at which point you can reduce the proof to:
```suggestion
  rw [← derivedSet_closure, isClosed_iff_derivedSet_subset]
  apply derivedSet_mono
  simp [← isClosed_iff_derivedSet_subset]
```

File: Mathlib/Topology/Perfect.lean [hunk 5]
```diff
@@ -168,6 +168,26 @@ theorem Perfect.splitting [T25Space α] (hC : Perfect C) (hnonempty : C.Nonempty
...
+    · simp
+    apply Set.union_subset_union_right
+    exact subset_closure
+  ) (Set.inter_singleton_nonempty.mpr hx) (by
+    obtain ⟨y, hy⟩ := Set.Nontrivial.exists_ne hu x
+    use y
+    simp only [Set.mem_inter_iff, hy, true_and]
+    apply subset_closure
+    simp [hy]
+  )
```
Comment (j-loreaux) [inline Mathlib/Topology/Perfect.lean:186]: the style is very strange here. I'm having trouble even making out what's going on. Can you improve it?

Comment (Command-Master) [reply Mathlib/Topology/Perfect.lean]: Is this better?

File: Mathlib/Topology/Perfect.lean [hunk 6]
```diff
@@ -172,18 +172,17 @@ lemma IsPreconnected.preperfect_of_nontrivial [T1Space α] {U : Set α} (hu : U.
...
+  · exact Set.inter_singleton_nonempty.mpr hx
+  · obtain ⟨y, hy⟩ := Set.Nontrivial.exists_ne hu x
     use y
     simp only [Set.mem_inter_iff, hy, true_and]
     apply subset_closure
     simp [hy]
-  )
   apply Set.Nonempty.right at h
   rw [Set.singleton_inter_nonempty, mem_closure_iff_clusterPt, ← acc_principal_iff_cluster] at h
   exact h
```
Comment (j-loreaux) [inline Mathlib/Topology/Perfect.lean:188]:
```suggestion
  · apply Set.Nonempty.right at h
    rw [Set.singleton_inter_nonempty, mem_closure_iff_clusterPt, ← acc_principal_iff_cluster] at h
    exact h
```

**Comment by j-loreaux** [comment]
bors d+

please complete the last change above, then wait for CI to finish before merging.

**Pull request closed without merging**